### PR TITLE
Add the bootloader code.

### PR DESCRIPTION
**Pull request by nomis52**
Tests are coming in the next PR.

### CODE REVIEW

File: Bootloader/firmware/src/bootloader.c [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+} TransferData;
+
+static TransferData g_transfer;
+
+/**
+ * @brief The buffer that holds the DFU Status response.
+ */
+static uint8_t g_status_response[GET_STATUS_RESPONSE_SIZE];
+
+/*
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.c]: **?

File: Bootloader/firmware/src/bootloader.c [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+  g_bootloader.usb_device = USB_DEVICE_HANDLE_INVALID;
+  g_bootloader.state = BOOTLOADER_STATE_INIT;
+  g_bootloader.dfu_state = DFU_STATE_IDLE;
+  g_bootloader.dfu_status = DFU_STATUS_OK;
+  g_bootloader.active_interface = DFU_ALT_INTERFACE_FIRMWARE;
+  PLIB_PORTS_PinDirectionOutputSet(PORTS_ID_0, LED_PORT_CHANNEL, LED_PORT_BIT);
+}
+
+void Bootloader_Tasks(void) {
+  // Flash the led to indicate we're in bootloader mode.
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.c]: SPaG LED

File: Bootloader/firmware/src/bootloader.c [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+    PLIB_PORTS_PinToggle(PORTS_ID_0, LED_PORT_CHANNEL, LED_PORT_BIT);
+    count = 0;
+  }
+
+  switch (g_bootloader.state) {
+    case BOOTLOADER_STATE_INIT:
+      g_bootloader.usb_device = USB_DEVICE_Open(USB_DEVICE_INDEX_0,
+                                                DRV_IO_INTENT_READWRITE);
+      if (g_bootloader.usb_device != USB_DEVICE_HANDLE_INVALID) {
+        // Register a callback with device layer to receive USB events.
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.c]: SPaG with the

File: Bootloader/firmware/src/bootloader.c [hunk 4]
```diff
@@ -0,0 +1,720 @@
...
+    case BOOTLOADER_STATE_WAIT_FOR_USB_CONFIGURATION:
+      // no op, waiting for USBEventHandler to change state.
+      break;
+
+    case BOOTLOADER_STATE_DFU:
+      if (g_bootloader.dfu_state == DFU_STATE_DNBUSY) {
+        ProcessDownload();
+      } else if (g_bootloader.dfu_state == DFU_STATE_MANIFEST_SYNC &&
+                 g_transfer.transfer_state == TRANSFER_LAST_BLOCK_RECEIVED) {
+        // The firmware may not be a multiple of 4, so write any remaining
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.c]: 4 bytes

File: Bootloader/firmware/src/bootloader.h [hunk 5]
```diff
@@ -13,30 +13,51 @@
...
+void Bootloader_Initialize(void);
 
 /**
  * @brief The bootloader event loop.
  */
-void APP_Tasks(void);
+void Bootloader_Tasks(void);
+
+/**
+ * @brief Return true if USB is active & configured.
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.h]: Check if USB...

File: Bootloader/firmware/src/bootloader.h [hunk 6]
```diff
@@ -13,30 +13,51 @@
...
+ * @returns true if USB is active & configured.
+ */
+bool Bootloader_USBActive();
+
+/**
+ * @brief Return true if USB is active & configured.
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.h]: Again

File: Bootloader/firmware/src/bootloader.h [hunk 7]
```diff
@@ -13,30 +13,51 @@
...
+ * @returns true if USB is active & configured.
+ */
+DFUState Bootloader_GetState();
+
+/**
+ * @brief Return true if USB is active & configured.
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.h]: Again

File: Bootloader/firmware/src/system_config/default/system_init.c [hunk 8]
```diff
@@ -221,20 +233,68 @@ const struct {
...
-sd002 = {
-  sizeof(sd002),
+PRODUCT_STRING_DESCRIPTOR = {
+  sizeof(PRODUCT_STRING_DESCRIPTOR),
   USB_DESCRIPTOR_STRING, {
     'J', 'a', ' ', 'R', 'u', 'l', 'e'
   }
 };
 
+/*
```
Comment (peternewman) [inline Bootloader/firmware/src/system_config/default/system_init.c]: Didn't we have this elsewhere?

File: Bootloader/firmware/src/system_config/default/system_init.c [hunk 9]
```diff
@@ -221,20 +233,68 @@ const struct {
...
  ***************************************/
-USB_DEVICE_STRING_DESCRIPTORS_TABLE stringDescriptors[3] = {
-  (const uint8_t * const) &sd000,
-  (const uint8_t * const) &sd001,
-  (const uint8_t * const) &sd002
+USB_DEVICE_STRING_DESCRIPTORS_TABLE stringDescriptors[6] = {
+  (const uint8_t * const) &LANGUAGE_STRING_DESCRIPTOR,
+  (const uint8_t * const) &MANUFACTURER_STRING_DESCRIPTOR,
+  (const uint8_t * const) &PRODUCT_STRING_DESCRIPTOR,
+  (const uint8_t * const) &serial_number_string_descriptor,
```
Comment (peternewman) [inline Bootloader/firmware/src/system_config/default/system_init.c]: Any reason this is lower case?

File: Bootloader/firmware/src/system_config/default/system_init.c [hunk 10]
```diff
@@ -192,8 +204,8 @@ const struct {
   uint16_t string[1];
 }
 
-sd000 = {
-  sizeof(sd000),
+LANGUAGE_STRING_DESCRIPTOR = {
+  sizeof(LANGUAGE_STRING_DESCRIPTOR),
   USB_DESCRIPTOR_STRING, {
     0x0409
```
Comment (peternewman) [inline Bootloader/firmware/src/system_config/default/system_init.c]: Is there not a table of constants for this?

**Comment by peternewman** [comment]
Generally LGTM

File: Bootloader/firmware/src/bootloader.c (hunk 4 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader.c]: Changed to 'The firmware size may not be ..."

File: Bootloader/firmware/src/bootloader.h (hunk 5 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader.h]: Done

File: Bootloader/firmware/src/bootloader.h (hunk 6 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader.h]: Done

File: Bootloader/firmware/src/bootloader.h (hunk 7 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader.h]: Done

File: Bootloader/firmware/src/system_config/default/system_init.c (hunk 10 above)
Comment (nomis52) [inline Bootloader/firmware/src/system_config/default/system_init.c]: Nope, it's all manual.

File: Bootloader/firmware/src/system_config/default/system_init.c (hunk 8 above)
Comment (nomis52) [inline Bootloader/firmware/src/system_config/default/system_init.c]: It exists in both the bootloader and main application as they appear as different USB devices. I want to clean up the USB descriptor code so I may be able to merge them later.

File: Bootloader/firmware/src/system_config/default/system_init.c (hunk 9 above)
Comment (nomis52) [inline Bootloader/firmware/src/system_config/default/system_init.c]: It's non-const.

File: Bootloader/firmware/src/bootloader.c (hunk 1 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader.c]: Fixed

File: Bootloader/firmware/src/bootloader.c (hunk 2 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader.c]: Fixed

File: Bootloader/firmware/src/bootloader.c (hunk 3 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader.c]: Fixed

File: Bootloader/firmware/src/system_config/default/system_init.c (hunk 8 above)
Comment (peternewman) [inline Bootloader/firmware/src/system_config/default/system_init.c]: Okay cool.

File: Bootloader/firmware/src/system_config/default/system_init.c (hunk 9 above)
Comment (peternewman) [inline Bootloader/firmware/src/system_config/default/system_init.c]: Ah understood.

File: Bootloader/firmware/src/system_config/default/system_init.c (hunk 10 above)
Comment (peternewman) [inline Bootloader/firmware/src/system_config/default/system_init.c]: I see

File: Bootloader/firmware/src/bootloader.c (hunk 1 above)
Comment (peternewman) [inline Bootloader/firmware/src/bootloader.c]: There were a few more like this too, not sure if they're intentional or not.

**Comment by peternewman** [comment]
LGTM

**Pull request merged**